File: 376.摆动序列.java
/*
 * @lc app=leetcode.cn id=376 lang=java
 *
 * [376] 摆动序列
 */

// @lc code=start
class Solution {
    public int wiggleMaxLength(int[] nums) {
        if (nums.length < 1) {
            return 0;
        }
        int lastDir = 0;
        int suc = 1;
        int start = 0;
        int lastValidValue = nums[0];
        for (start = 1; start < nums.length; start++) {
            if (nums[start] != nums[start - 1]) {
                lastDir = nums[start] > nums[start - 1] ? 1 : -1;//斜率
                lastValidValue = nums[start];
                suc=suc+1;
                break;
            }
        }


        for (; start < nums.length; start++) {
            if (-1 * lastDir * (nums[start] - nums[start - 1]) > 0) {
                suc++;
                lastDir = -1 * lastDir;
            } lastValidValue = nums[start];
        }
        return suc;
    }
}
// @lc code=end

File: 376.摆动序列.py
#
# @lc app=leetcode.cn id=376 lang=python3
#
# [376] 摆动序列
#

# @lc code=start
class Solution:
    def wiggleMaxLength(self, nums: List[int]) -> int:
        n = len(nums)
        if n <= 1: return n
        residual = [0] * (n-1)
        for i in range(n-1):
            residual[i] = nums[i+1] - nums[i]
        res = n
        for i in range(n-2):
            if residual[i] * residual[i+1] >= 0:
                res -= 1
        if residual[-1]==0: res -= 1
        return res
# @lc code=end

